WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -352,27 +376,60 @@ public IEnumerable<Money> Decompose(IEnumerable<Money> myInputCoinEffectiveValue
 	/// <summary>
 	/// Greedily decomposes an amount to the given denominations.
 	/// </summary>
-	private IEnumerable<ulong> BreakDown(Money coininputEffectiveValue, IEnumerable<ulong> denominations)
+	private IEnumerable<Output> BreakDown(Money coininputEffectiveValue, IEnumerable<Output> denominations)
 	{
 		var remaining = coininputEffectiveValue;
 
-		foreach (var denomPlusFee in denominations)
+		List<Output> denoms = new();
+
+		foreach (var denom in denominations)
 		{
-			if (denomPlusFee < MinAllowedOutputAmountPlusFee || remaining < MinAllowedOutputAmountPlusFee)
+			if (denom.Amount < MinAllowedOutputAmount || remaining < MinAllowedOutputAmount + ChangeFee)
 			{
 				break;
 			}
 
-			while (denomPlusFee <= remaining)
+			while (denom.EffectiveCost <= remaining)
 			{
-				yield return denomPlusFee;
-				remaining -= denomPlusFee;
+				denoms.Add(denom);
+				remaining -= denom.EffectiveCost;
 			}
 		}
 
-		if (remaining >= MinAllowedOutputAmountPlusFee)
+		if (remaining >= MinAllowedOutputAmount + ChangeFee)
+		{
+			denoms.Add(Output.FromAmount(remaining, ChangeScriptType, FeeRate));
+		}
+
+		return denoms;
+	}
+
+	public class Output
```

Comment:
This class should implement `GetHashCode`  because it is used as a key in a dictionary, theorically speaking. However the UT pass consistenty. I don't know if dotnet is smarter now but I would provide an implementation anyway.

Reply:
I was reading a bit more and yes, or you provide a GetHashCode method or, you have to specify the IEqualityComparer, otherwise this works just by chance.

Reply:
Added